storey/sources.py

```diff
@@ -619,56 +617,87 @@ def __init__(
+                        return single_key, key_field
+                    key.append(single_key)
+            else:
+                key = body[self._key_field]
+                if pd.isna(key):
+                    return key, self._key_field
+        return key, None
 
     async def _run_loop(self):
         for df in self._dfs:
-            for namedtuple in df.itertuples():
-                create_event = True
-                body = namedtuple._asdict()
-                index = body.pop("Index")
-                if len(df.index.names) > 1:
-                    for i, index_column in enumerate(df.index.names):
-                        body[index_column] = index[i]
-                elif df.index.names[0] is not None:
-                    body[df.index.names[0]] = index
-
-                key = None
-                if self._key_field:
-                    if isinstance(self._key_field, list):
-                        key = []
-                        for key_field in self._key_field:
-                            if key_field not in body or pandas.isna(body[key_field]):
-                                create_event = False
-                                break
-                            key.append(body[key_field])
-                    else:
-                        key = body[self._key_field]
-                        if key is None:
-                            create_event = False
-                if create_event:
+            columns = list(df.columns)
+            if not df.index.empty and not (len(df.index.names) == 1 and df.index.names[0] is None):
+                df = df.reset_index(drop=False)
+            for body in df.to_dict("records"):
+                key, none_key_column = self._get_key(body=body)
+                if not none_key_column:
                     if self._id_field:
-                        id = body[self._id_field]
+                        line_id = body[self._id_field]
                     else:
-                        id = self._get_uuid()
-                    event = Event(body, key=key, id=id)
+                        line_id = self._get_uuid()
+                    element = self._get_element(body=body, columns=columns)
+                    event = Event(element, key=key, id=line_id)
                     await self._do_downstream(event)
                 else:
                     if self.context:
-                        self.context.logger.error(f"For {body} value of key {key_field} is None")
+                        self.context.logger.error(f"value of key {none_key_column} is {key} For {body}")
         return await self._do_downstream(_termination_obj)
 
+    def _validate_fields(self, df, path=""):
+        path_message = f" File path: {path}." if path else ""
+        df = df.reset_index()
+        if self._key_field:
+            key_fields = [self._key_field] if not isinstance(self._key_field, list) else self._key_field
+            missing_keys = list(set(key_fields) - set(df.columns))
```

Comment:
I think it better be something like
```suggestion
missing_keys = [key_field for key_field in key_fields if key_field in df.columns]
```
Because sets are unordered and may mess up the order of the keys and confuse the user.

Reply:
* not in